paimon-core/src/main/java/org/apache/paimon/manifest/ManifestFileMeta.java

```diff
@@ -283,16 +283,11 @@ public static Optional<List<ManifestFileMeta>> tryFullCompaction(
                 Predicate predicate = predicateOpt.get();
                 FieldStatsArraySerializer fieldStatsArraySerializer =
                         new FieldStatsArraySerializer(partitionType);
-                for (; j < base.size(); j++) {
-                    // TODO: optimize this to binary search.
-                    ManifestFileMeta file = base.get(j);
-                    if (predicate.test(
-                            file.numAddedFiles + file.numDeletedFiles,
-                            file.partitionStats.fields(fieldStatsArraySerializer))) {
-                        break;
-                    } else {
-                        result.add(file);
-                    }
+                int searchResult =
+                        binarySearchWithFilter(
+                                base, 0, base.size() - 1, predicate, fieldStatsArraySerializer);
```

Comment:
low should not be 0, low should be `j`.

We need to more tests to cover here. The reason for not using binary search at the beginning was that there was not enough energy to write enough tests to cover it.